### ISSUE TITLE
Segfault on program exit when openssl is statically linked

### ISSUE THREAD

**Issue by pasweistorz** [openssl/openssl#17040]
We face an issue that might be similar to #15898. We have an application that uses both openssl 1.1.1j and Qt 5.15.2. If openssl is statically linked both into our application and into Qt and if Qt is dynamically linked to our application we get a segmentation fault at program exit. As in #15898 we also see two calls of `ossl_init_register_atexit_ossl_`.

I tried to create a minimal working example without Qt that has similar behavior. Here, `liba` plays the role of Qt. The linker flag `-Bsymbolic-functions` seems to be important for reproducing this issue. It is used by Qt when the shared Qt libraries are linked so I added it here, too, as we have no influence on it.

This example program doesn't crash but rather seems to hang forever instead at program exit. It is built with CMake 3.16.3 and gcc 9.3.0 on Ubuntu 20.04 with openssl 1.1.1f.

CMakeLists.txt:

```
cmake_minimum_required(VERSION 3.16)

project(repro)

set(OPENSSL_USE_STATIC_LIBS TRUE)
find_package(OpenSSL REQUIRED)

add_library(a SHARED a.cpp)
target_link_options(a PRIVATE LINKER:-Bsymbolic-functions)
target_link_libraries(a PRIVATE ${OPENSSL_SSL_LIBRARY} ${OPENSSL_CRYPTO_LIBRARY})

add_executable(repro main.cpp)
target_link_libraries(repro PUBLIC ${OPENSSL_SSL_LIBRARY} ${OPENSSL_CRYPTO_LIBRARY} -lpthread -ldl a)
```

main.cpp:

```
#include "openssl/x509.h"

void a();

int main()
{
    a();

    X509* x509 = X509_new();
    X509_free(x509);
}
```

a.cpp:

```
#include <iostream>
#include "openssl/rand.h"
#include "openssl/ssl.h"

void a()
{
    if (OPENSSL_init_ssl(0, nullptr) != 1)
    {
        std::cout << "OPENSSL_init_ssl failed" << std::endl;
        return;
    }

    if (!RAND_status())
    {
        std::cout << "RAND_status failed" << std::endl;
        return;
    }

    std::cout << "a successful" << std::endl;
}
```

A first analysis:

In our minimal working example the `-Bsymbolic-functions` linker flag has the effect that the global symbol `rand_meth` (which is part of the application and `liba` because openssl is statically linked into both of them) is bound to the definition within the application and shared between the application and `liba` whereas references to global function symbols within `liba` are bound to the definitions within `liba` and, thus, aren't shared with the application.

We can see this in the call stack of the application.

Breakpoint at `ossl_init_register_atexit_ossl_`

First hit:

```
#0  0x00007ffff7d8b1d0 in ossl_init_register_atexit_ossl_ () from /bug-repro/build/liba.so
#1  0x00007ffff7fa147f in __pthread_once_slow (once_control=0x7ffff7f8759c <register_atexit>, init_routine=0x7ffff7d8b1d0 <ossl_init_register_atexit_ossl_>) at pthread_once.c:116
#2  0x00007ffff7dd8cbd in CRYPTO_THREAD_run_once () from /bug-repro/build/liba.so
#3  0x00007ffff7d8b643 in OPENSSL_init_crypto () from /bug-repro/build/liba.so
#4  0x00007ffff7ce4505 in OPENSSL_init_ssl () from /bug-repro/build/liba.so
#5  0x00007ffff7ce4100 in a () at /bug-repro/a.cpp:7
#6  0x00005555555ca12a in main () at /bug-repro/main.cpp:7
```

Second hit:

```
#0  0x00005555555f22c0 in ossl_init_register_atexit_ossl_ ()
#1  0x00007ffff7fa147f in __pthread_once_slow (once_control=0x5555557f367c <register_atexit>, init_routine=0x5555555f22c0 <ossl_init_register_atexit_ossl_>) at pthread_once.c:116
#2  0x0000555555603f2d in CRYPTO_THREAD_run_once ()
#3  0x00005555555f2733 in OPENSSL_init_crypto ()
#4  0x00005555555f9561 in do_rand_drbg_init_ossl_ ()
#5  0x00007ffff7fa147f in __pthread_once_slow (once_control=0x5555557f3738 <rand_drbg_init>, init_routine=0x5555555f9550 <do_rand_drbg_init_ossl_>) at pthread_once.c:116
#6  0x0000555555603f2d in CRYPTO_THREAD_run_once ()
#7  0x00005555555f970b in RAND_DRBG_get0_master ()
#8  0x00005555555f9743 in drbg_status ()
#9  0x00007ffff7ce413c in a () at /bug-repro/a.cpp:13
#10 0x00005555555ca12a in main () at /bug-repro/main.cpp:7
```

As `drbg_status` is called from `RAND_STATUS` via function pointer `status` located in global and shared symbol `rand_meth` we switch to the openssl "instance" within the application. Thus, `ossl_init_register_at_exit_ossl_` is called a second time, although we only refer to libcrypto from `liba` and not from the application. Registering the same exit function twice might lead to the strange behavior at program exit.

I'm not sure if this is an openssl bug at all. Nevertheless I think if the symbol `rand_meth` wasn't exported the problem wouldn't occur anymore. Hopefully my comments also help in making progress regarding #15898.

**Comment by mattcaswell** [comment]
This does indeed look very similar to #15898.

I was able to reproduce this using your simple reproducer.

Ultimately the problem is that OpenSSL is being linked in twice and there is confusion between the two.

The `-Bsymbolic-functions` linker symbols has this description:

> –B symbolic | symbolic-functions
These options are only available when creating a shared object. The –B symbolic option causes global symbol references to be bound to definitions within the object being built, if a definition exists. The –B symbolic-functions restricts this binding to function symbols.
>
> Normally, references to global symbols within a shared object are not bound until runtime, regardless of whether a definition of the symbol exists within the shared object. This model allows a definition of the same symbol from an executable, or from another shared object, to override the object's own definition.

So, when compiling liba, global symbol references to *functions* are being bound to the object being built, i.e. OpenSSL functions symbols must come from liba's version of OpenSSL. Global *variables* however are *not* bound in this way and (according to the last paragraph) are "not bound until runtime" and come from the executable's version of OpenSSL.

This is significant for two global variables in use here: `rand_meth` and `global_engine_lock`.

`rand_meth` contains a structure holding function pointers to OpenSSL's default rand method. Since `rand_meth` is a global variable, when liba uses it, it ends up using the executables copy which contains function pointers to the executable's version of the rand functions. These functions end up calling the executable's version of `OPENSSL_init_crypto` which registers an `atexit` handler.

Separate to this liba explicitly calls liba's version of `OPENSSL_init_ssl` which calls liba's version of  `OPENSSL_init_crypto`. This *also* registers an `atexit` handler.

Normally we only ever register one `atexit` handler since this is controlled by a "run_once" variable - which is supposed to mean the atexit registration code only ever runs once. However the "run_once" variable is not a global - it is declared "static", so liba and the executable both have their own separate copies.

The atexit handler is responsible for freeing up global resources. When it gets called the first time, it briefly uses the `global_engine_lock`, and then when its finished, it frees it up and sets it to NULL. 

The second time the atexit handler is called it again attempts to briefly use the `global_engine_lock`, but this time its been set to NULL and a NULL deref occurs - resulting in a seg fault.

It seems to me that this is ultimately caused by the use of `symbolic-functions` resulting in some things coming from liba's copy of OpenSSL, and some things coming from the executable's copy. If this option was omitted then (I assume) everything would come from the executables copy, and everything should work. Similarly if "-Bsymbolic" was used instead then *everything* liba uses should come from liba's copy, and *everything* that the executable uses should come from *its* copy. So, this should also work. I tested using "-Bsymbolic" instead in your reproducer and it did seem to work without crashing. I also tried not linking OpenSSL to the executable:

```` diff
$ diff -u CMakeLists.txt.orig CMakeLists.txt
--- CMakeLists.txt.orig	2021-11-15 16:22:24.956309336 +0000
+++ CMakeLists.txt	2021-11-15 16:22:48.956154278 +0000
@@ -12,4 +12,4 @@
 target_link_libraries(a PRIVATE ${OPENSSL_SSL_LIBRARY} ${OPENSSL_CRYPTO_LIBRARY})
 
 add_executable(repro main.cpp)
-target_link_libraries(repro PUBLIC ${OPENSSL_SSL_LIBRARY} ${OPENSSL_CRYPTO_LIBRARY} -lpthread -ldl a)
+target_link_libraries(repro PUBLIC -lpthread -ldl a)
````

This also seems to work.

I do not consider this to be a bug in OpenSSL.
